apollo-gradle-plugin/src/main/kotlin/com/apollographql/apollo/gradle/internal/SchemaDownloader.kt

```diff
@@ -39,30 +44,56 @@ object SchemaDownloader {
             addHeader(it.key, it.value)
           }
         }
-        .url(endpoint)
+        .url(url)
         .build()
 
-    val response = OkHttpClient.Builder()
-        .connectTimeout(connectTimeoutSeconds, TimeUnit.SECONDS)
-        .readTimeout(readTimeoutSeconds, TimeUnit.SECONDS)
-        .build()
-        .newCall(request).execute()
+    val response = newOkHttpClient()
+        .newCall(request)
+        .execute()
 
     if (!response.isSuccessful) {
       throw Exception("cannot get schema: ${response.code}:\n${response.body?.string()}")
     }
 
-    schema.parentFile?.mkdirs()
+    return response
+  }
+
+  fun downloadIntrospection(
+      endpoint: String,
+      headers: Map<String, String>
+  ): String {
+
+    val response = executeQuery(introspectionQuery, endpoint, headers)
+
+    return response.body.use { responseBody ->
+      responseBody!!.string()
+    }
+  }
 
-    response.body.use { responseBody ->
-      if (schema.extension.toLowerCase() == "json") {
-        schema.writeText(responseBody!!.string())
-      } else {
-        IntrospectionSchema(responseBody!!.byteStream()).toSDL(schema)
+  fun downloadRegistry(graph: String, key: String, variant: String): String? {
+    val registryQuery = """
+    query {
```

Comment:
Could we make sure to name the operation here? Additionally, using variables is a best practice, along with identifying your client traffic using the `apollographql-client-name` header.

Maybe something like this?

```
query DownloadSchema($graphID: ID!, $variant: String! = "current") {
  service(id: $graphID) {
    variant(name: $variant) {
      activeSchemaPublish {
        schema {
          document
        }
      }
    }
  }
}
```